tests/integration/test_charm.py

```diff
@@ -68,3 +75,85 @@ async def test_owasp_modsecurity_custom_rules(app_url_modsec_ignore: str):
     """
     response = requests.get(app_url_modsec_ignore)
     assert response.status_code == 200
+
+
+async def test_owasp_modsecurity_crs_relation(ops_test: OpsTest, app_name: str, tmp_path: Path):
+    """
+    arrange: given charm has been built and deployed
+    act: relate the ingress integrator with any-charm and use any-charm to toggle modsecurity
+        option through relation via ingress lib.
+    assert: modsecurity annotations should be attached and detached from the kubernetes ingress
+        resource according to the modsecurity option.
+    """
+    await ops_test.model.applications[app_name].set_config(
+        {"owasp-modsecurity-crs": "false", "owasp-modsecurity-custom-rules": ""}
+    )
+    any_charm_tmp_path = tmp_path / "any-charm"
+    if not any_charm_tmp_path.exists():
+        subprocess.run(  # nosec
+            ["git", "clone", "https://github.com/canonical/any-charm.git", any_charm_tmp_path]
+        )
+
+    any_charm = await ops_test.build_charm(any_charm_tmp_path)
+    ingress_lib = Path("lib/charms/nginx_ingress_integrator/v0/ingress.py").read_text()
+    any_charm_src_overwrite = {
+        "ingress.py": ingress_lib,
+        "any_charm.py": dedent(
+            """\
+        from ingress import IngressRequires
+        from any_charm_base import AnyCharmBase
+        class AnyCharm(AnyCharmBase):
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+                self.ingress = IngressRequires(
+                    self,
+                    {
+                        "service-hostname": "any",
+                        "service-name": self.app.name,
+                        "service-port": 80,
+                        "owasp-modsecurity-crs": True
```

Comment:
One question here, which could be some miscomprehension on my side:
The charm uses the method `_get_config_or_relation_data` to get the value which will be use to insert in the ingress annontation
This `_get_config_or_relation_data` method will take the `nginx-ingress-integrator` charm config if it exists, or the value defined on the related charm (`any-charm` in our case here)
When I lookg at `config.yaml` of `ngninx-ingress-integrator` there's a default value to `false` for `owasp-modsecurity-crs` config...
So in my understanding what ever value is defined by the relation, since there's a config value to `false`, it will will be `false`

What do I miss ?

Reply:
The `owasp-modsecurity-crs` is not in `BOOLEAN_CONFIG_FIELDS`, so `_get_config` will return `None` if the `owasp-modsecurity-crs` set to `False`. So if `owasp-modsecurity-crs` config is `False` the modsecurity will depends on the relation. Is this a bug or a feature?

The more concerning part is `rewrite-enabled` which is out of this PR's scope. But `rewrite-enabled` is in the `BOOLEAN_CONFIG_FIELDS`, which makes it impossible to update that option via relation.

Reply:
That's indeed what I missed. Well honestly I feel like this overriding rules "config overrides relation" or "relation overrides config" depending on the parameter are very confusing, and could lead to unexpected behavior from user perspective
I'm wondering if the charm shouldn't be much more opinionated and just forbid relation overriding the configuration, recent discussion with Tom and Arturo led to the conclusion that in most use cases there is 1 deploy ment of the nignx-ingress operator related to 1 backend charm (and not multiple backend related to it). Therefore there won't be consequences to only allow changes from the config side

Reply:
Yes, but it will break the backward compatibility, a major one. The charm rewrite would not be very difficult though.